File: emoji-library/src/main/java/com/blackhat/emoji/helper/EmojiconsPopup.java
package com.blackhat.emoji.helper;

import android.app.Activity;
import android.content.Context;
import android.graphics.Color;
import android.graphics.Rect;
import android.os.Build;
import android.os.Handler;
import android.os.SystemClock;
import android.util.DisplayMetrics;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.view.ViewTreeObserver.OnGlobalLayoutListener;
import android.view.WindowManager;
import android.view.WindowManager.LayoutParams;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.PopupWindow;

import androidx.viewpager.widget.PagerAdapter;
import androidx.viewpager.widget.ViewPager;

import java.util.Arrays;
import java.util.List;

import com.blackhat.emoji.R;
import com.blackhat.emoji.emoji.Cars;
import com.blackhat.emoji.emoji.Electr;
import com.blackhat.emoji.emoji.Emojicon;
import com.blackhat.emoji.emoji.Food;
import com.blackhat.emoji.emoji.Nature;
import com.blackhat.emoji.emoji.People;
import com.blackhat.emoji.emoji.Sport;
import com.blackhat.emoji.emoji.Symbols;

public class EmojiconsPopup extends PopupWindow implements ViewPager.OnPageChangeListener, EmojiconRecents {
    private int mEmojiTabLastSelectedIndex = -1;
    private View[] mEmojiTabs;
    private PagerAdapter mEmojisAdapter;
    private EmojiconRecentsManager mRecentsManager;
    private int keyBoardHeight = 0;
    private Boolean pendingOpen = false;
    private Boolean isOpened = false;
    public EmojiconGridView.OnEmojiconClickedListener onEmojiconClickedListener;
    OnEmojiconBackspaceClickedListener onEmojiconBackspaceClickedListener;
    OnSoftKeyboardOpenCloseListener onSoftKeyboardOpenCloseListener;
    View rootView;
    Context mContext;
    boolean mUseSystemDefault = false;
    View view;
    int positionPager = 0;
    boolean setColor = false;
    String iconPressedColor = "#495C66";
    String tabsColor = "#DCE1E2";
    String backgroundColor = "#E6EBEF";

    private ViewPager emojisPager;

    public EmojiconsPopup(View rootView, Context mContext, boolean useSystemDefault, String iconPressedColor, String tabsColor, String backgroundColor) {
        super(mContext);
        this.setColor = true;
        this.backgroundColor = backgroundColor;
        this.iconPressedColor = iconPressedColor;
        this.tabsColor = tabsColor;
        this.mUseSystemDefault = useSystemDefault;
        this.mContext = mContext;
        this.rootView = rootView;
        View customView = createCustomView();
        setContentView(customView);
        setSoftInputMode(LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE);
        setSize(LayoutParams.MATCH_PARENT, 255);
        setBackgroundDrawable(null);


    }

    public EmojiconsPopup(View rootView, Context mContext, boolean useSystemDefault) {
        super(mContext);
        this.mUseSystemDefault = useSystemDefault;
        this.mContext = mContext;
        this.rootView = rootView;
        View customView = createCustomView();
        setContentView(customView);
        setSoftInputMode(LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE);
        setSize(LayoutParams.MATCH_PARENT, 255);
        setBackgroundDrawable(null);

    }

    public void setOnSoftKeyboardOpenCloseListener(OnSoftKeyboardOpenCloseListener listener) {
        this.onSoftKeyboardOpenCloseListener = listener;
    }

    public void setOnEmojiconClickedListener(EmojiconGridView.OnEmojiconClickedListener listener) {
        this.onEmojiconClickedListener = listener;
    }

    public void setOnEmojiconBackspaceClickedListener(OnEmojiconBackspaceClickedListener listener) {
        this.onEmojiconBackspaceClickedListener = listener;
    }

    public void showAtBottom() {
        showAtLocation(rootView, Gravity.BOTTOM, 0, 0);
    }

    public void showAtBottomPending() {
        if (isKeyBoardOpen())
            showAtBottom();
        else
            pendingOpen = true;
    }

    public Boolean isKeyBoardOpen() {
        return isOpened;
    }

    @Override
    public void dismiss() {
        super.dismiss();
        EmojiconRecentsManager
                .getInstance(mContext).saveRecents();
    }

    public void setSizeForSoftKeyboard() {
        rootView.getViewTreeObserver().addOnGlobalLayoutListener(new OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                Rect r = new Rect();
                rootView.getWindowVisibleDisplayFrame(r);

                int screenHeight = getUsableScreenHeight();
                int heightDifference = screenHeight
                        - (r.bottom - r.top);
                int resourceId = mContext.getResources()
                        .getIdentifier("status_bar_height",
                                "dimen", "android");
                if (resourceId > 0) {
                    heightDifference -= mContext.getResources()
                            .getDimensionPixelSize(resourceId);
                }
                if (heightDifference > 100) {
                    keyBoardHeight = heightDifference;
                    setSize(LayoutParams.MATCH_PARENT, keyBoardHeight);
                    if (isOpened == false) {
                        if (onSoftKeyboardOpenCloseListener != null)
                            onSoftKeyboardOpenCloseListener.onKeyboardOpen(keyBoardHeight);
                    }
                    isOpened = true;
                    if (pendingOpen) {
                        showAtBottom();
                        pendingOpen = false;
                    }
                } else {
                    isOpened = false;
                    if (onSoftKeyboardOpenCloseListener != null)
                        onSoftKeyboardOpenCloseListener.onKeyboardClose();
                }
            }
        });
    }

    private int getUsableScreenHeight() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            DisplayMetrics metrics = new DisplayMetrics();

            WindowManager windowManager = (WindowManager) mContext.getSystemService(Context.WINDOW_SERVICE);
            windowManager.getDefaultDisplay().getMetrics(metrics);

            return metrics.heightPixels;

        } else {
            return rootView.getRootView().getHeight();
        }
    }

    public void setSize(int width, int height) {
        setWidth(width);
        setHeight(height);
    }

    public void updateUseSystemDefault(boolean mUseSystemDefault) {
        if (view != null) {
            mEmojisAdapter = null;
            positionPager = emojisPager.getCurrentItem();
            dismiss();

            this.mUseSystemDefault = mUseSystemDefault;
            setContentView(createCustomView());
            //mEmojisAdapter.notifyDataSetChanged();
            mEmojiTabs[positionPager].setSelected(true);
            emojisPager.setCurrentItem(positionPager);
            onPageSelected(positionPager);
            if (!isShowing()) {

                //If keyboard is visible, simply show the emoji popup
                if (isKeyBoardOpen()) {
                    showAtBottom();
                    // changeEmojiKeyboardIcon(emojiButton, R.drawable.ic_action_keyboard);
                }

                //else, open the text keyboard first and immediately after that show the emoji popup
                else {
                    showAtBottomPending();
                    // changeEmojiKeyboardIcon(emojiButton, R.drawable.ic_action_keyboard);
                }
            }


        }
    }


    private View createCustomView() {
        LayoutInflater inflater = (LayoutInflater) mContext.getSystemService(Activity.LAYOUT_INFLATER_SERVICE);
        view = inflater.inflate(R.layout.emojicons, null, false);
        emojisPager = view.findViewById(R.id.emojis_pager);
        LinearLayout tabs = view.findViewById(R.id.emojis_tab);

        emojisPager.setOnPageChangeListener(this);
        EmojiconRecents recents = this;
        mEmojisAdapter = new EmojisPagerAdapter(
                Arrays.asList(
                        new EmojiconRecentsGridView(mContext, null, null, this, mUseSystemDefault),
                        new EmojiconGridView(mContext, People.DATA, recents, this, mUseSystemDefault),
                        new EmojiconGridView(mContext, Nature.DATA, recents, this, mUseSystemDefault),
                        new EmojiconGridView(mContext, Food.DATA, recents, this, mUseSystemDefault),
                        new EmojiconGridView(mContext, Sport.DATA, recents, this, mUseSystemDefault),
                        new EmojiconGridView(mContext, Cars.DATA, recents, this, mUseSystemDefault),
                        new EmojiconGridView(mContext, Electr.DATA, recents, this, mUseSystemDefault),
                        new EmojiconGridView(mContext, Symbols.DATA, recents, this, mUseSystemDefault)

                )
        );
        emojisPager.setAdapter(mEmojisAdapter);
        mEmojiTabs = new View[8];

        mEmojiTabs[0] = view.findViewById(R.id.emojis_tab_0_recents);
        mEmojiTabs[1] = view.findViewById(R.id.emojis_tab_1_people);
        mEmojiTabs[2] = view.findViewById(R.id.emojis_tab_2_nature);
        mEmojiTabs[3] = view.findViewById(R.id.emojis_tab_3_food);
        mEmojiTabs[4] = view.findViewById(R.id.emojis_tab_4_sport);
        mEmojiTabs[5] = view.findViewById(R.id.emojis_tab_5_cars);
        mEmojiTabs[6] = view.findViewById(R.id.emojis_tab_6_elec);
        mEmojiTabs[7] = view.findViewById(R.id.emojis_tab_7_sym);
        for (int i = 0; i < mEmojiTabs.length; i++) {
            final int position = i;
            mEmojiTabs[i].setOnClickListener(new OnClickListener() {
                @Override
                public void onClick(View v) {
                    emojisPager.setCurrentItem(position);
                }
            });
        }


        emojisPager.setBackgroundColor(Color.parseColor(backgroundColor));
        tabs.setBackgroundColor(Color.parseColor(tabsColor));
        for (int x = 0; x < mEmojiTabs.length; x++) {
            ImageButton btn = (ImageButton) mEmojiTabs[x];
            btn.setColorFilter(Color.parseColor(iconPressedColor));
        }

        ImageButton imgBtn = (ImageButton) view.findViewById(R.id.emojis_backspace);
        imgBtn.setColorFilter(Color.parseColor(iconPressedColor));
        imgBtn.setBackgroundColor(Color.parseColor(backgroundColor));


        view.findViewById(R.id.emojis_backspace).setOnTouchListener(new RepeatListener(500, 50, new OnClickListener() {

            @Override
            public void onClick(View v) {
                if (onEmojiconBackspaceClickedListener != null)
                    onEmojiconBackspaceClickedListener.onEmojiconBackspaceClicked(v);
            }
        }));

        // get last selected page
        mRecentsManager = EmojiconRecentsManager.getInstance(view.getContext());
        int page = mRecentsManager.getRecentPage();
        // last page was recents, check if there are recents to use
        // if none was found, go to page 1
        if (page == 0 && mRecentsManager.size() == 0) {
            page = 1;
        }

        if (page == 0) {
            onPageSelected(page);
        } else {
            emojisPager.setCurrentItem(page, false);
        }
        return view;
    }

    @Override
    public void addRecentEmoji(Context context, Emojicon emojicon) {
        EmojiconRecentsGridView fragment = ((EmojisPagerAdapter) emojisPager.getAdapter()).getRecentFragment();
        fragment.addRecentEmoji(context, emojicon);
    }


    @Override
    public void onPageScrolled(int i, float v, int i2) {
    }

    @Override
    public void onPageSelected(int i) {
        if (mEmojiTabLastSelectedIndex == i) {
            return;
        }
        switch (i) {
            case 0:
            case 1:
            case 2:
            case 3:
            case 4:
            case 5:
            case 6:
            case 7:

                if (mEmojiTabLastSelectedIndex >= 0 && mEmojiTabLastSelectedIndex < mEmojiTabs.length) {
                    mEmojiTabs[mEmojiTabLastSelectedIndex].setSelected(false);
                }
                mEmojiTabs[i].setSelected(true);
                mEmojiTabLastSelectedIndex = i;
                mRecentsManager.setRecentPage(i);
                break;
        }
    }

    @Override
    public void onPageScrollStateChanged(int i) {
    }

    private static class EmojisPagerAdapter extends PagerAdapter {
        private List<EmojiconGridView> views;

        public EmojiconRecentsGridView getRecentFragment() {
            for (EmojiconGridView it : views) {
                if (it instanceof EmojiconRecentsGridView)
                    return (EmojiconRecentsGridView) it;
            }
            return null;
        }

        public EmojisPagerAdapter(List<EmojiconGridView> views) {
            super();
            this.views = views;
        }

        @Override
        public int getCount() {
            return views.size();
        }


        @Override
        public Object instantiateItem(ViewGroup container, int position) {
            View v = views.get(position).rootView;
            ((ViewPager) container).addView(v, 0);
            return v;
        }

        @Override
        public void destroyItem(ViewGroup container, int position, Object view) {
            ((ViewPager) container).removeView((View) view);
        }

        @Override
        public boolean isViewFromObject(View view, Object key) {
            return key == view;
        }
    }

    public static class RepeatListener implements View.OnTouchListener {

        private Handler handler = new Handler();

        private int initialInterval;
        private final int normalInterval;
        private final OnClickListener clickListener;

        private Runnable handlerRunnable = new Runnable() {
            @Override
            public void run() {
                if (downView == null) {
                    return;
                }
                handler.removeCallbacksAndMessages(downView);
                handler.postAtTime(this, downView, SystemClock.uptimeMillis() + normalInterval);
                clickListener.onClick(downView);
            }
        };

        private View downView;

        public RepeatListener(int initialInterval, int normalInterval, OnClickListener clickListener) {
            if (clickListener == null)
                throw new IllegalArgumentException("null runnable");
            if (initialInterval < 0 || normalInterval < 0)
                throw new IllegalArgumentException("negative interval");

            this.initialInterval = initialInterval;
            this.normalInterval = normalInterval;
            this.clickListener = clickListener;
        }

        public boolean onTouch(View view, MotionEvent motionEvent) {
            switch (motionEvent.getAction()) {
                case MotionEvent.ACTION_DOWN:
                    downView = view;
                    handler.removeCallbacks(handlerRunnable);
                    handler.postAtTime(handlerRunnable, downView, SystemClock.uptimeMillis() + initialInterval);
                    clickListener.onClick(view);
                    return true;
                case MotionEvent.ACTION_UP:
                case MotionEvent.ACTION_CANCEL:
                case MotionEvent.ACTION_OUTSIDE:
                    handler.removeCallbacksAndMessages(downView);
                    downView = null;
                    return true;
            }
            return false;
        }
    }

    public interface OnEmojiconBackspaceClickedListener {
        void onEmojiconBackspaceClicked(View v);
    }

    public interface OnSoftKeyboardOpenCloseListener {
        void onKeyboardOpen(int keyBoardHeight);

        void onKeyboardClose();
    }
}

File: emoji-library/src/main/java/com/blackhat/emoji/helper/EmojiconsPopup.kt
package com.blackhat.emoji.helper

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.graphics.Color
import android.graphics.Rect
import android.os.Handler
import android.os.SystemClock
import android.util.DisplayMetrics
import android.view.Gravity
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.View.OnClickListener
import android.view.ViewGroup
import android.view.WindowManager
import android.view.WindowManager.LayoutParams
import android.widget.ImageButton
import android.widget.LinearLayout
import android.widget.PopupWindow

import androidx.viewpager.widget.PagerAdapter
import androidx.viewpager.widget.ViewPager

import java.util.Arrays
import java.util.Objects

import com.blackhat.emoji.R
import com.blackhat.emoji.emoji.Cars
import com.blackhat.emoji.emoji.Electr
import com.blackhat.emoji.emoji.Emojicon
import com.blackhat.emoji.emoji.Food
import com.blackhat.emoji.emoji.Nature
import com.blackhat.emoji.emoji.People
import com.blackhat.emoji.emoji.Sport
import com.blackhat.emoji.emoji.Symbols

class EmojiconsPopup : PopupWindow, ViewPager.OnPageChangeListener, EmojiconRecents {
    private var mEmojiTabLastSelectedIndex = -1
    private lateinit var mEmojiTabs: Array<View>
    private var mEmojisAdapter: PagerAdapter? = null
    private var mRecentsManager: EmojiconRecentsManager? = null
    private var keyBoardHeight = 0
    private var pendingOpen: Boolean? = false
    var isKeyBoardOpen: Boolean? = false
        private set
    var onEmojiconClickedListener: EmojiconGridView.OnEmojiconClickedListener? = null

    private var onEmojiconBackspaceClickedListener: OnEmojiconBackspaceClickedListener? = null
    private var onSoftKeyboardOpenCloseListener: OnSoftKeyboardOpenCloseListener? = null
    private var rootView: View? = null
    private var mContext: Context? = null
    private var mUseSystemDefault: Boolean = false
    private var view: View? = null
    private var iconPressedColor = "#495C66"
    private var tabsColor = "#DCE1E2"
    private var backgroundColor = "#E6EBEF"

    private var emojisPager: ViewPager? = null

    private val usableScreenHeight: Int
        get() {
            val metrics = DisplayMetrics()

            val windowManager = mContext!!.getSystemService(Context.WINDOW_SERVICE) as WindowManager
            Objects.requireNonNull(windowManager).defaultDisplay.getMetrics(metrics)

            return metrics.heightPixels

        }

    constructor(rootView: View, mContext: Context, useSystemDefault: Boolean, iconPressedColor: String, tabsColor: String, backgroundColor: String) : super(mContext) {
        this.backgroundColor = backgroundColor
        this.iconPressedColor = iconPressedColor
        this.tabsColor = tabsColor
        this.mUseSystemDefault = useSystemDefault
        this.mContext = mContext
        this.rootView = rootView
        val customView = createCustomView()
        contentView = customView
        softInputMode = LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE
        setSize(255)
        setBackgroundDrawable(null)


    }

    constructor(rootView: View, mContext: Context, useSystemDefault: Boolean) : super(mContext) {
        this.mUseSystemDefault = useSystemDefault
        this.mContext = mContext
        this.rootView = rootView
        val customView = createCustomView()
        contentView = customView
        softInputMode = LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE
        setSize(255)
        setBackgroundDrawable(null)

    }

    fun setOnSoftKeyboardOpenCloseListener(listener: OnSoftKeyboardOpenCloseListener) {
        this.onSoftKeyboardOpenCloseListener = listener
    }

    fun setOnEmojiconBackspaceClickedListener(listener: OnEmojiconBackspaceClickedListener) {
        this.onEmojiconBackspaceClickedListener = listener
    }

    fun showAtBottom() {
        showAtLocation(rootView, Gravity.BOTTOM, 0, 0)
    }

    fun showAtBottomPending() {
        if (isKeyBoardOpen!!)
            showAtBottom()
        else
            pendingOpen = true
    }

    override fun dismiss() {
        super.dismiss()
        EmojiconRecentsManager
                .getInstance(mContext!!).saveRecents()
    }

    fun setSizeForSoftKeyboard() {
        rootView!!.viewTreeObserver.addOnGlobalLayoutListener {
            val r = Rect()
            rootView!!.getWindowVisibleDisplayFrame(r)

            val screenHeight = usableScreenHeight
            var heightDifference = screenHeight - (r.bottom - r.top)
            val resourceId = mContext!!.resources
                    .getIdentifier("status_bar_height",
                            "dimen", "android")
            if (resourceId > 0) {
                heightDifference -= mContext!!.resources
                        .getDimensionPixelSize(resourceId)
            }
            if (heightDifference > 100) {
                keyBoardHeight = heightDifference
                setSize(keyBoardHeight)
                if ((!isKeyBoardOpen!!)) {
                    if (onSoftKeyboardOpenCloseListener != null)
                        onSoftKeyboardOpenCloseListener!!.onKeyboardOpen(keyBoardHeight)
                }
                isKeyBoardOpen = true
                if (pendingOpen!!) {
                    showAtBottom()
                    pendingOpen = false
                }
            } else {
                isKeyBoardOpen = false
                if (onSoftKeyboardOpenCloseListener != null)
                    onSoftKeyboardOpenCloseListener!!.onKeyboardClose()
            }
        }
    }

    private fun setSize(height: Int) {
        width = ViewGroup.LayoutParams.MATCH_PARENT
        setHeight(height)
    }

    fun updateUseSystemDefault(mUseSystemDefault: Boolean) {
        if (view != null) {
            mEmojisAdapter = null
            val positionPager = emojisPager!!.currentItem
            dismiss()

            this.mUseSystemDefault = mUseSystemDefault
            contentView = createCustomView()
            //mEmojisAdapter.notifyDataSetChanged();
            mEmojiTabs[positionPager].isSelected = true
            emojisPager!!.currentItem = positionPager
            onPageSelected(positionPager)
            if (!isShowing) {

                //If keyboard is visible, simply show the emoji popup
                if (isKeyBoardOpen!!) {
                    showAtBottom()
                    // changeEmojiKeyboardIcon(emojiButton, R.drawable.ic_action_keyboard);
                } else {
                    showAtBottomPending()
                    // changeEmojiKeyboardIcon(emojiButton, R.drawable.ic_action_keyboard);
                }//else, open the text keyboard first and immediately after that show the emoji popup
            }


        }
    }


    @SuppressLint("InflateParams")
    private fun createCustomView(): View {
        val inflater = mContext!!.getSystemService(Activity.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        view = Objects.requireNonNull(inflater).inflate(R.layout.emojicons, null, false)
        emojisPager = view!!.findViewById(R.id.emojis_pager)
        val tabs = view!!.findViewById<LinearLayout>(R.id.emojis_tab)

        emojisPager!!.addOnPageChangeListener(this)
        mEmojisAdapter = EmojisPagerAdapter(
                Arrays.asList(
                        EmojiconRecentsGridView(mContext!!, null, this, this, mUseSystemDefault),
                        EmojiconGridView(mContext!!, People.DATA, this, this, mUseSystemDefault),
                        EmojiconGridView(mContext!!, Nature.DATA, this, this, mUseSystemDefault),
                        EmojiconGridView(mContext!!, Food.DATA, this, this, mUseSystemDefault),
                        EmojiconGridView(mContext!!, Sport.DATA, this, this, mUseSystemDefault),
                        EmojiconGridView(mContext!!, Cars.DATA, this, this, mUseSystemDefault),
                        EmojiconGridView(mContext!!, Electr.DATA, this, this, mUseSystemDefault),
                        EmojiconGridView(mContext!!, Symbols.DATA, this, this, mUseSystemDefault)

                )
        )
        emojisPager!!.adapter = mEmojisAdapter
        mEmojiTabs = arrayOf(
                view!!.findViewById(R.id.emojis_tab_0_recents),
                view!!.findViewById(R.id.emojis_tab_1_people),
                view!!.findViewById(R.id.emojis_tab_2_nature),
                view!!.findViewById(R.id.emojis_tab_3_food),
                view!!.findViewById(R.id.emojis_tab_4_sport),
                view!!.findViewById(R.id.emojis_tab_5_cars),
                view!!.findViewById(R.id.emojis_tab_6_elec),
                view!!.findViewById(R.id.emojis_tab_7_sym)
        )

//        mEmojiTabs[0] = view!!.findViewById(R.id.emojis_tab_0_recents)
//        mEmojiTabs[1] = view!!.findViewById(R.id.emojis_tab_1_people)
//        mEmojiTabs[2] = view!!.findViewById(R.id.emojis_tab_2_nature)
//        mEmojiTabs[3] = view!!.findViewById(R.id.emojis_tab_3_food)
//        mEmojiTabs[4] = view!!.findViewById(R.id.emojis_tab_4_sport)
//        mEmojiTabs[5] = view!!.findViewById(R.id.emojis_tab_5_cars)
//        mEmojiTabs[6] = view!!.findViewById(R.id.emojis_tab_6_elec)
//        mEmojiTabs[7] = view!!.findViewById(R.id.emojis_tab_7_sym)
        for (i in mEmojiTabs.indices) {
            mEmojiTabs[i].setOnClickListener { v -> emojisPager!!.currentItem = i }
        }


        emojisPager!!.setBackgroundColor(Color.parseColor(backgroundColor))
        tabs.setBackgroundColor(Color.parseColor(tabsColor))
        for (mEmojiTab in mEmojiTabs!!) {
            val btn = mEmojiTab as ImageButton
            btn.setColorFilter(Color.parseColor(iconPressedColor))
        }

        val imgBtn = view!!.findViewById<ImageButton>(R.id.emojis_backspace)
        imgBtn.setColorFilter(Color.parseColor(iconPressedColor))
        imgBtn.setBackgroundColor(Color.parseColor(backgroundColor))


        view!!.findViewById<View>(R.id.emojis_backspace).setOnTouchListener(RepeatListener(500, 50, OnClickListener { v ->
            if (onEmojiconBackspaceClickedListener != null)
                v?.let { onEmojiconBackspaceClickedListener!!.onEmojiconBackspaceClicked(it) }
        }))

        // get last selected page
        mRecentsManager = EmojiconRecentsManager.getInstance(view!!.context)
        var page = mRecentsManager!!.recentPage
        // last page was recents, check if there are recents to use
        // if none was found, go to page 1
        if (page == 0 && mRecentsManager!!.size == 0) {
            page = 1
        }

        if (page == 0) {
            onPageSelected(page)
        } else {
            emojisPager!!.setCurrentItem(page, false)
        }
        return view as View
    }

    override fun addRecentEmoji(context: Context, emojicon: Emojicon) {
        val fragment = (Objects.requireNonNull<PagerAdapter>(emojisPager!!.adapter) as EmojisPagerAdapter).recentFragment
        fragment!!.addRecentEmoji(context, emojicon)
    }


    override fun onPageScrolled(i: Int, v: Float, i2: Int) {}

    override fun onPageSelected(i: Int) {
        if (mEmojiTabLastSelectedIndex == i) {
            return
        }
        when (i) {
            0, 1, 2, 3, 4, 5, 6, 7 -> {

                if (mEmojiTabLastSelectedIndex >= 0 && mEmojiTabLastSelectedIndex < mEmojiTabs!!.size) {
                    mEmojiTabs!![mEmojiTabLastSelectedIndex].isSelected = false
                }
                mEmojiTabs!![i].isSelected = true
                mEmojiTabLastSelectedIndex = i
                mRecentsManager!!.recentPage = i
            }
        }
    }

    override fun onPageScrollStateChanged(i: Int) {}

    private class EmojisPagerAdapter internal constructor(private val views: List<EmojiconGridView>) : PagerAdapter() {

        internal val recentFragment: EmojiconRecentsGridView?
            get() {
                for (it in views) {
                    if (it is EmojiconRecentsGridView)
                        return it
                }
                return null
            }

        override fun getCount(): Int {
            return views.size
        }


        override fun instantiateItem(container: ViewGroup, position: Int): Any {
            val v = views[position].rootView
            container.addView(v, 0)
            return v
        }

        override fun destroyItem(container: ViewGroup, position: Int, view: Any) {
            container.removeView(view as View)
        }

        override fun isViewFromObject(view: View, key: Any): Boolean {
            return key === view
        }
    }

    class RepeatListener internal constructor(private val initialInterval: Int, private val normalInterval: Int, private val clickListener: OnClickListener?) : View.OnTouchListener {

        private val handler = Handler()

        private val handlerRunnable = object : Runnable {
            override fun run() {
                if (downView == null) {
                    return
                }
                handler.removeCallbacksAndMessages(downView)
                handler.postAtTime(this, downView, SystemClock.uptimeMillis() + normalInterval)
                clickListener?.onClick(downView)
            }
        }

        private var downView: View? = null

        init {
            requireNotNull(clickListener) { "null runnable" }
            require(!(initialInterval < 0 || normalInterval < 0)) { "negative interval" }
        }

        @SuppressLint("ClickableViewAccessibility")
        override fun onTouch(view: View, motionEvent: MotionEvent): Boolean {
            when (motionEvent.action) {
                MotionEvent.ACTION_DOWN -> {
                    downView = view
                    handler.removeCallbacks(handlerRunnable)
                    handler.postAtTime(handlerRunnable, downView, SystemClock.uptimeMillis() + initialInterval)
                    clickListener!!.onClick(view)
                    return true
                }
                MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL, MotionEvent.ACTION_OUTSIDE -> {
                    handler.removeCallbacksAndMessages(downView)
                    downView = null
                    return true
                }
            }
            return false
        }
    }

    interface OnEmojiconBackspaceClickedListener {
        fun onEmojiconBackspaceClicked(v: View)
    }

    interface OnSoftKeyboardOpenCloseListener {
        fun onKeyboardOpen(keyBoardHeight: Int)

        fun onKeyboardClose()
    }
}